CRM/Upgrade/Incremental/php/FiveSixtyOne.php

```diff
@@ -46,6 +46,38 @@ public function setPreUpgradeMessage(&$preUpgradeMessage, $rev, $currentVer = NU
    */
   public function upgrade_5_61_alpha1($rev): void {
     $this->addTask(ts('Upgrade DB to %1: SQL', [1 => $rev]), 'runSql', $rev);
+    $this->addTask(ts('Dedupe cache table'), 'dedupeCache');
+    $this->addTask(ts('Drop index %1', [1 => 'civicrm_cache.UI_group_path_date']), 'dropIndex', 'civicrm_cache', 'UI_group_path_date');
+    $this->addTask(ts('Create index %1', [1 => 'civicrm_cache.UI_group_name_path']), 'addIndex', 'civicrm_cache', [['group_name', 'path']], 'UI');
+  }
+
+  /**
+   * Remove extraneous/duplicate records from `civicrm_cache`.
+   *
+   * Formally, the cache table allowed multiple (key,value) pairs if created at different times.
+   * In practice, this cleanup should generally do nothing -- the `SqlGroup::set()` has had duplicate
+   * prevention, and the cache will flush at the end of the upgrade anyway. Never-the-less, if
+   * duplicates are somehow in there, then we should cleanly remove them rather than let the upgrade fail.
+   *
+   * @param \CRM_Queue_TaskContext $ctx
+   * @return bool
+   * @throws \Civi\Core\Exception\DBQueryException
+   */
+  public static function dedupeCache($ctx): bool {
+    $duplicates = CRM_Core_DAO::executeQuery('
+      SELECT c.id FROM civicrm_cache c
+      LEFT JOIN (SELECT group_name, path, max(created_date) newest FROM civicrm_cache GROUP BY group_name, path) recent
```

Comment:
Hmm - you didn't think about just truncating the cache?

Reply:
Yeah, I briefly considered that there are several formulations that would probably do the job but didn't linger on it. I'll grant this formulation is more code-heavy than others, but it's one-off, and - behaviorally speaking - it's the lightest formulation (i.e. less likely to have some unforeseen interaction). Overall, it's probably 6/half-dozen.

FWIW, I did do some `r-run` with redundant records to ensure that it got the right ones.